chrome/content/zotero/import/folderImport.js

```diff
@@ -0,0 +1,241 @@
+		const fileDataByHash = {};
+		files.forEach((fileData, index) => {
+			const hash = fileHashes[index];
+			if (hash in fileDataByHash) {
+				fileDataByHash[hash].parentCollectionIDs.push(...fileData.parentCollectionIDs);
+			}
+			else {
+				fileDataByHash[hash] = fileData;
+			}
+		});
+
+		// advance progress to account for duplicates found within file structure
+		// these files won't be imported nor recognized so advance 2 ticks per file
+		this._progress += 2 * (files.length - Object.keys(fileDataByHash).length);
+		this._itemDone();
+
+		const attachmentItemHashLookup = {};
+		const attachmentItems = await Promise.all(Object.entries(fileDataByHash).map(
+			async ([hash, { name, path, parentCollectionIDs, mimeType }]) => {
+				const options = {
+					collections: parentCollectionIDs,
+					contentType: mimeType,
+					file: path,
+					libraryID,
+				};
+
+				let attachmentItem = null;
+				
+				if ((this.types.includes(mimeType) || multimatch(name, this.fileTypes, { nocase: true }).length > 0)) {
+					const existingItem = await findItemByHash(libraryID, hash);
+
+					if (existingItem) {
+						existingItem.setCollections([...existingItem.getCollections(), ...parentCollectionIDs]);
+						existingItem.saveTx({ skipSelect: true });
+					}
+					else {
+						if (linkFiles) {
+							attachmentItem = await Zotero.Attachments.linkFromFile(options);
+						}
+						else {
+							attachmentItem = await Zotero.Attachments.importFromFile(options);
+						}
+						
+						this.newItems.push(attachmentItem);
+						attachmentItemHashLookup[attachmentItem.id] = hash;
+					}
+				}
+
+				if (attachmentItem && !Zotero.RecognizePDF.canRecognize(attachmentItem)) {
+					attachmentItem.setRelations({ 'zotero:fileHash': hash });
+					await attachmentItem.saveTx({ skipSelect: true });
+					attachmentItem = null;
+				}
+				this._progress++;
+				this._itemDone();
+				return attachmentItem;
+			}
+		));
+
+
+		// discard unrecognizable items, increase progress for discarded items
+		const recognizableItems = attachmentItems.filter(item => item !== null);
+		this._progress += attachmentItems.length - recognizableItems.length;
+		this._itemDone();
+
+		const recognizeQueue = Zotero.ProgressQueues.get('recognize');
+		const itemsToSavePostRecognize = [];
+		
+		const processRecognizedItem = ({ status, id }) => {
+			const updatedItem = recognizableItems.find(i => i.id === id);
+			if (status === Zotero.ProgressQueue.ROW_SUCCEEDED) {
+				const recognizedItem = updatedItem.parentItem;
+				if (recognizedItem && id in attachmentItemHashLookup) {
+					recognizedItem.setRelations({ 'zotero:attachmentHash': attachmentItemHashLookup[id] });
+					itemsToSavePostRecognize.push(recognizedItem);
+				}
+			}
+			if (status === Zotero.ProgressQueue.ROW_FAILED) {
+				if (updatedItem && id in attachmentItemHashLookup) {
+					updatedItem.setRelations({ 'zotero:fileHash': attachmentItemHashLookup[id] });
```

Comment:
Generally, these are used to avoid duplicates on re-import, but it's been a while, and I can't remember right why we need both. I'll see if we can get away with just fileHash and come back to you.

Reply:
OK, thinking about this, I think we do want to track a file hash (and a last-check timestamp) for all attachments so that we can use it for duplicate detection/notification, including matching on existing items before metadata retrieval (and for cases where metadata retrieval doesn't work).

Probably easier if I do that part — and it will need to be added to the dataserver too, and reconciled with https://github.com/zotero/zotero/pull/2854 — but maybe we should just disable folder import in the dialog for now so that we can merge this and get a Z7 beta out? We can take out the `fileHash`/`attachmentHash` parts to avoid future confusion and just add a `throw` to make sure nobody runs the folder-import code in the time being.

Reply:
(I'm writing up a plan for the hash, but we'll definitely want to merge the rest of this first.)

Reply:
I've disabled folder import and related tests. This option won't appear in the import wizard and `translate()` throws an error just in case. All the code is still there, except for the part that checked/stored hashes.